File: src/python/xerxes/channels/adapters/whatsapp.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""WhatsApp channel adapter.

Targets the WhatsApp Business Cloud API on Meta's Graph endpoint. Parses
inbound webhook payloads (entry → changes → value → messages) and sends
``text`` messages through ``POST /<phone_number_id>/messages``. Meta's
``hub.verify_token`` handshake is the operator's responsibility — the
adapter only handles the JSON event format.
"""

from __future__ import annotations

import typing as tp

from .._helpers import WebhookChannel, http_post, parse_json_body
from ..types import ChannelMessage, MessageDirection


class WhatsAppChannel(WebhookChannel):
    """WhatsApp Business Cloud API (Meta Graph) adapter."""

    name = "whatsapp"

    def __init__(
        self,
        access_token: str,
        phone_number_id: str,
        *,
        http_client: tp.Any = None,
        api_version: str = "v18.0",
    ) -> None:
        """Build the channel.

        Args:
            access_token: Meta Graph API access token.
            phone_number_id: WhatsApp Business phone-number id; the leading
                path segment of every outbound API call.
            http_client: Optional HTTP client override forwarded to
                ``http_post``.
            api_version: Meta Graph API version. Defaults to ``"v18.0"``.
        """
        super().__init__()
        self.access_token = access_token
        self.phone_number_id = phone_number_id
        self.api_version = api_version
        self._http = http_client

    def _parse_inbound(self, headers, body):
        """Walk the WhatsApp webhook envelope and emit one message per entry.

        The Cloud API can batch many messages into one webhook: each
        ``entry`` may have multiple ``changes``, and each change's ``value``
        may have multiple ``messages``. We unpack the full tree.

        Args:
            headers: HTTP headers (unused).
            body: Raw JSON webhook body.

        Returns:
            Every parsed user message in delivery order; empty if the
            envelope contained none.
        """
        data = parse_json_body(body)
        out = []
        for entry in data.get("entry", []):
            for change in entry.get("changes", []):
                value = change.get("value") or {}
                for msg in value.get("messages", []):
                    text = (msg.get("text") or {}).get("body", "") or msg.get("button", {}).get("text", "")
                    out.append(
                        ChannelMessage(
                            text=text,
                            channel=self.name,
                            channel_user_id=msg.get("from", ""),
                            room_id=msg.get("from", ""),
                            platform_message_id=msg.get("id", ""),
                            direction=MessageDirection.INBOUND,
                        )
                    )
        return out

    async def _send_outbound(self, message):
        """Send a ``text`` message via ``POST /<phone_number_id>/messages``.

        Args:
            message: Outbound message. ``room_id`` or ``channel_user_id``
                becomes the recipient phone number (E.164 without ``+``);
                ``text`` carries the body.
        """
        url = f"https://graph.facebook.com/{self.api_version}/{self.phone_number_id}/messages"
        body = {
            "messaging_product": "whatsapp",
            "to": message.room_id or message.channel_user_id,
            "type": "text",
            "text": {"body": message.text},
        }
        http_post(
            url,
            json_body=body,
            headers={"Authorization": f"Bearer {self.access_token}"},
            http_client=self._http,
        )

File: xerxes/src/channels/whatsApp.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { createHmac } from 'node:crypto'

import { postJson, providerUrl, type ChannelFetch } from './http.js'
import {
  arrayValue,
  outboundDestination,
  recordValue,
  requiredOption,
  stringValue,
  type RelayChannelTransport,
} from './relay.js'
import { createChannelMessage, MessageDirection, type ChannelMessage } from './types.js'
import { constantTimeEqualStrings, webhookHeaderValue } from './webhookSignatures.js'
import { parseJsonBody, WebhookChannel, type WebhookHeaders, type WebhookResponse } from './webhooks.js'

const WHATSAPP_GRAPH_API = 'https://graph.facebook.com/'
const WHATSAPP_SIGNATURE_PREFIX = 'sha256='
/** Warned exactly once per adapter when inbound signatures are not verified. */
const UNVERIFIED_SIGNATURE_WARNING = 'WhatsApp webhook signature verification is DISABLED: set the'
  + ' "app_secret" channel option so X-Hub-Signature-256 payloads can be validated; unverified'
  + ' webhooks must not face a public network.'

/** WhatsApp Cloud API webhook relay with Graph HTTP sends and optional signature checks. */
export const WHATSAPP_TRANSPORT: RelayChannelTransport = {
  inbound: 'webhook-relay',
  outbound: 'http-api',
  unsupported: ['persistent WhatsApp socket transport'],
}

export interface WhatsAppChannelOptions {
  readonly accessToken: string
  /**
   * Optional Meta app secret. When set, inbound webhooks must carry a valid
   * `X-Hub-Signature-256` header (hex HMAC-SHA256 of the raw body) or they
   * are rejected; when omitted, signatures are accepted unverified and the
   * adapter warns loudly once.
   */
  readonly appSecret?: string
  readonly apiBaseUrl?: string
  readonly apiVersion?: string
  readonly fetchImplementation?: ChannelFetch
  readonly phoneNumberId: string
}

/**
 * WhatsApp Business Cloud API webhook relay and text sender.
 *
 * The HTTP edge must answer Meta's GET verification challenge;
 * `whatsAppWebhookChallenge` is provided for that. POST deliveries are
 * verified against `X-Hub-Signature-256` whenever an `appSecret` is
 * configured.
 */
export class WhatsAppChannel extends WebhookChannel {
  readonly name = 'whatsapp'
  readonly transport = WHATSAPP_TRANSPORT

  private readonly accessToken: string
  private readonly apiBaseUrl: string
  private readonly apiVersion: string
  private readonly fetchImplementation: ChannelFetch | undefined
  private readonly phoneNumberId: string
  private readonly appSecret: string
  private warnedUnverifiedSignatures = false

  constructor(options: WhatsAppChannelOptions) {
    super()
    this.accessToken = requiredOption(options.accessToken, 'WhatsApp accessToken')
    this.apiBaseUrl = options.apiBaseUrl ?? WHATSAPP_GRAPH_API
    this.apiVersion = options.apiVersion ?? 'v23.0'
    this.fetchImplementation = options.fetchImplementation
    this.phoneNumberId = requiredOption(options.phoneNumberId, 'WhatsApp phoneNumberId')
    this.appSecret = options.appSecret?.trim() ?? ''
  }

  override async handleWebhook(headers: WebhookHeaders, body: Uint8Array): Promise<WebhookResponse> {
    if (!this.appSecret) {
      this.warnUnverifiedSignaturesOnce()
    } else if (!whatsAppSignatureMatches(headers, body, this.appSecret)) {
      // Fail closed: a configured app secret makes valid signatures mandatory.
      return { status: 401, body: 'unauthorized' }
    }
    return super.handleWebhook(headers, body)
  }

  protected parseInbound(
    _headers: WebhookHeaders,
    body: Uint8Array,
  ): readonly ChannelMessage[] {
    const payload = parseJsonBody(body)
    const messages: ChannelMessage[] = []
    for (const entryValue of arrayValue(payload.entry)) {
      const entry = recordValue(entryValue)
      for (const changeValue of arrayValue(entry.changes)) {
        const change = recordValue(changeValue)
        const value = recordValue(change.value)
        for (const incomingValue of arrayValue(value.messages)) {
          const incoming = recordValue(incomingValue)
          const text = whatsAppText(incoming)
          if (!text) {
            continue
          }
          const sender = stringValue(incoming.from)
          messages.push(createChannelMessage({
            channel: this.name,
            direction: MessageDirection.INBOUND,
            text,
            channelUserId: sender,
            roomId: sender,
            platformMessageId: stringValue(incoming.id),
            metadata: { type: stringValue(incoming.type) },
          }))
        }
      }
    }
    return messages
  }

  protected async sendOutbound(message: ChannelMessage): Promise<void> {
    const recipient = outboundDestination(message, 'WhatsApp')
    await postJson(providerUrl(this.apiBaseUrl, `${this.apiVersion}/${this.phoneNumberId}/messages`), {
      body: {
        messaging_product: 'whatsapp',
        to: recipient,
        type: 'text',
        text: { body: message.text },
      },
      headers: { Authorization: `Bearer ${this.accessToken}` },
      ...(this.fetchImplementation ? { fetchImplementation: this.fetchImplementation } : {}),
    })
  }

  private warnUnverifiedSignaturesOnce(): void {
    if (this.warnedUnverifiedSignatures) return
    this.warnedUnverifiedSignatures = true
    console.warn(UNVERIFIED_SIGNATURE_WARNING)
  }
}

/** Validate Meta's `X-Hub-Signature-256`: hex HMAC-SHA256 of the raw body. */
function whatsAppSignatureMatches(headers: WebhookHeaders, body: Uint8Array, appSecret: string): boolean {
  const provided = webhookHeaderValue(headers, 'x-hub-signature-256')
  if (!provided || !provided.startsWith(WHATSAPP_SIGNATURE_PREFIX)) return false
  const expected = WHATSAPP_SIGNATURE_PREFIX + createHmac('sha256', appSecret).update(body).digest('hex')
  return constantTimeEqualStrings(provided, expected)
}

/** Return Meta's verification challenge only when its configured token matches. */
export function whatsAppWebhookChallenge(
  query: Readonly<Record<string, string | undefined>>,
  verifyToken: string,
): string | undefined {
  if (query['hub.mode'] !== 'subscribe' || query['hub.verify_token'] !== verifyToken) {
    return undefined
  }
  return query['hub.challenge']
}

function whatsAppText(message: Readonly<Record<string, unknown>>): string {
  return stringValue(recordValue(message.text).body)
    || stringValue(recordValue(message.button).text)
    || stringValue(recordValue(recordValue(message.interactive).button_reply).title)
    || stringValue(recordValue(recordValue(message.interactive).list_reply).title)
}
